lib/srv/app/session.go

```diff
@@ -286,51 +288,25 @@ func (s *Server) newStreamWriter(chunkID string) (events.StreamWriter, error) {
 		return nil, trace.Wrap(err)
 	}
 
-	// Create a sync or async streamer depending on configuration of cluster.
-	streamer, err := s.newStreamer(chunkID, recConfig)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	streamWriter, err := events.NewAuditWriter(events.AuditWriterConfig{
-		// Audit stream is using server context, not session context,
-		// to make sure that session is uploaded even after it is closed
-		Context:      s.closeContext,
-		Streamer:     streamer,
-		Clock:        s.c.Clock,
-		SessionID:    rsession.ID(chunkID),
-		Namespace:    apidefaults.Namespace,
-		ServerID:     s.c.HostID,
-		RecordOutput: recConfig.GetMode() != types.RecordOff,
-		Component:    teleport.ComponentApp,
-		ClusterName:  clusterName.GetClusterName(),
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	return streamWriter, nil
-}
-
-// newStreamer returns sync or async streamer based on the configuration
-// of the server and the session, sync streamer sends the events
-// directly to the auth server and blocks if the events can not be received,
-// async streamer buffers the events to disk and uploads the events later
-func (s *Server) newStreamer(chunkID string, recConfig types.SessionRecordingConfig) (events.Streamer, error) {
-	if services.IsRecordSync(recConfig.GetMode()) {
-		s.log.Debugf("Using sync streamer for session chunk %v.", chunkID)
-		return s.c.AuthClient, nil
+	cfg := events.SessionWriterConfig{
+		Context:     ctx,
+		Clock:       s.c.Clock,
+		SessionID:   rsession.ID(chunkID),
+		Namespace:   apidefaults.Namespace,
+		ServerID:    s.c.HostID,
+		Component:   teleport.ComponentApp,
+		ClusterName: clusterName.GetClusterName(),
 	}
-
-	s.log.Debugf("Using async streamer for session chunk %v.", chunkID)
 	uploadDir := filepath.Join(
 		s.c.DataDir, teleport.LogsDir, teleport.ComponentUpload,
 		events.StreamingSessionsDir, apidefaults.Namespace,
 	)
-	fileStreamer, err := filesessions.NewStreamer(uploadDir)
+	streamWriter, err := recorder.New(recConfig, cfg, uploadDir, s.c.AuthClient)
```

Comment:
previous implementation using  `events.NewTeeStreamer` avoids sending `AppSessionRequestEvent` events to audit event (`AppSessionRequestEvent` only goes to session). I believe that logic now should be adjusted in `common.Audit`

Reply:
Thanks and good catch, fixed